docs/source/examples/checkpointer.rst

```diff
@@ -0,0 +1,319 @@
+            }
+
+        Recipe State:
+            {
+                "optimizer": ...,
+                "epoch": ...,
+                ...
+            }
+
+To restart from a previous checkpoint file, you'll need to make the following changes
+to the config file
+
+.. code-block:: yaml
+
+    checkpointer:
+
+        # checkpointer to use
+        _component_: torchtune.utils.FullModelHFCheckpointer
+
+        # directory with the checkpoint files
+        # this should match the output_dir above
+        checkpoint_dir: <checkpoint_dir>
+
+        # checkpoint files. These refer to intermediate checkpoints
+        # and will always have a ".pt" extensions
+        checkpoint_files: [
+            hf_model_0001_0.pt,
+            hf_model_0002_0.pt,
+        ]
+
+        # if we're restarting a previous run, we need to specify
+        # the file with the checkpoint state
+        recipe_checkpoint: recipe_state.pt
+
+        # dir for saving the output checkpoints. Usually set
+        # to be the same as checkpoint_dir
+        output_dir: <checkpoint_dir>
+
+        # model_type which specifies how to convert the state_dict
+        # into a format which TorchTune understands
+        model_type: LLAMA2
+
+    # set to True if restarting training
+    resume_from_checkpoint: True
+
+
+Checkpointing for LoRA
+----------------------
+
+In TorchTune, we output both the adapter weights and the full model "merged" weights
+for LoRA. The "merged" checkpoint can be used just like you would use the source
+checkpoint with any post-training tools.
+
+The primary difference between the two use cases is when you want to resume training
+from a checkpoint. In this case, the checkpointer needs access to both the initial frozen
+base model weights as well as the learnt adapter weights. The config for this scenario
+looks something like this:
+
+
+.. code-block:: yaml
+
+    checkpointer:
+
+        # checkpointer to use
+        _component_: torchtune.utils.FullModelHFCheckpointer
+
+        # directory with the checkpoint files
+        # this should match the output_dir above
+        checkpoint_dir: <checkpoint_dir>
+
+        # checkpoint files. This is the ORIGINAL frozen checkpoint
+        # and NOT the merged checkpoint output during training
+        checkpoint_files: [
+            pytorch_model-00001-of-00002.bin,
+            pytorch_model-00002-of-00002.bin,
+        ]
+
+        # this refers to the adapter weights learnt during training
+        adapter_checkpoint: adapter_0.pt
+
+        # the file with the checkpoint state
+        recipe_checkpoint: recipe_state.pt
+
+        # dir for saving the output checkpoints. Usually set
+        # to be the same as checkpoint_dir
+        output_dir: <checkpoint_dir>
+
+        # model_type which specifies how to convert the state_dict
+        # into a format which TorchTune understands
+        model_type: LLAMA2
+
+    # set to True if restarting training
+    resume_from_checkpoint: True
```

Comment:
I feel like this ends kinda abruptly. Would either add some bit about offramps or even just a generic conclusion to wrap things up. Given that you're ending with LoRA, an example of generation with the merged checkpoint in e.g. HF format could be a nice way to tie everything together.

Reply:
Good point!